.github/workflows/provenance.yaml

```diff
@@ -54,7 +54,7 @@ jobs:
           ./scripts/build_provenance \
           "oak_functions" \
           ./target/x86_64-unknown-linux-musl/release/oak_functions_loader_base \
```

Comment:
Is this name still valid? How many binaries are generated now? We probably need to update `./scripts/build_provenance` to generate multiple provenances. No?

Reply:
`build-functions-server-variants` currently builds the two binaries for the two variants `base` and `unsafe`.

If intrpret lines 54-56 correctly, it seems we did not have a provenance for `unsafe` before. Do we want one now?

Reply:
You are right. That path is still valid. I don't think we need a provenance for the `unsafe` version. At least not now. But if there is going to be a `bare-metal` version in the future, we'd probably need that. But maybe this change is not related to that. (I don't quite have the context for this change 🙂)

Reply:
@rbehjati  BTW once we add TOML files for each "transparently released" binary, we should consider adding one for the unsafe version too. In fact, since they are both generated by the same command now, that would be a nice use case to show that we can track provenance of multiple artifacts generated from the same process (i.e. perhaps a single TOML file will specify both binaries as output, to match the current logic). We'll see when we get there anyways

Reply:
I think, from the perspective of generating provenances, it is better to have the possibility of generating binaries separately. 

We can have a single provenance referencing multiple subjects, but in this particular case, the two binaries have very different security properties, and while it is possible to have separate claims for them, having a human user (e.g., an auditor) constantly look for `oak_functions_loader_base` vs `oak_functions_loader_unsafe` is not good usability.

Reply:
I think it makes sense to have a single command that generates a number of related binaries or libraries. But having a single command for generating a number of **alternative** binaries with opposing security properties, and including them all in a single provenance sounds like a bad design to me, as it can increase the chance of mistakes.

Reply:
Ah, sorry, merged this already. Shall we open an issue together?

Reply:
Can we have the previous `build_functions_server`, but with a new `all` variant (perhaps as default) to generate all variants as it now is in `build_functions_server_variants`? Also, I still don't know why we need to generate all variants together :)

Reply:
Building binaries both simplifies the interface of `xtask`, because the server variant can be set (and conflict) in two places: in (i) `example.toml` and (ii) with a command line flag. 

Now if `build_function_server_variants` always builds both variants we can remove (ii).

But: I was only thinking from the perspective of *calling* the binary (in an example), where I need to specify the variant anyway---and by building both before I know whatever I am picking is up-to-date.

Now I am also not sure if I fully understand the problem with building them in one go for provenance?

Reply:
> the server variant can be set (and conflict) in two places: in (i) example.toml and (ii) with a command line flag

Do you mean the `run-functions-example` command has a `server-variant` flag? 

> Now I am also not sure if I fully understand the problem with building them in one go for provenance?

I think including both `base` and `unsafe` variants in the subject of a single provenance file is wrong, as I explained above. We could use `build_function_server_variants` in the provenance, and include only the `base` variant in the subject of the provenance. But then the provenance could be considered misleading, because the command is actually generating an `unsafe` version too. The command could potentially be generating other malicious files too, which the provenance does not say anything about. I don't think we have a way of documenting in the provenance that "the command generates additional binaries that have similar functionality, but very different security properties than the subject of the provenance". Moreover, from the perspective of generating provenances, the choice of using a command that produces a bunch of other binaries too sounds very random. I think keeping the previous flags, and adding an `all` variant is a simple solution that can solve both problems with little effort.

Reply:
A couple of points:

- what I suggested was just a single TOML file for two expected outputs, but each output would still have its own provenance file; I think for our use cases we never expect more than one output in a provenance file; anyways I agree it is probably not even worth doing this with the TOML file only, it may just be confusing
- that said, I think it's fine to have a single command generate multiple binaries. In fact, any command probably generates thousands of files in the process (e.g. intermediate artifacts in the target dir, cache entries, etc.), which is why the TOML file needs to know which one is the actual output file to measure.

Reply:
Ah, I thought the idea was to use a single TOML file and generate a single provenance containing both variants in the subject.

> any command probably generates thousands of files in the process (e.g. intermediate artifacts in the target dir, cache entries, etc.),

True. I still *feel* like it is better to use a command that generates the fewest number of files, and ideally only the ones that are required for generating the file in the subject.